### PR TITLE
WIP: Add audit log module

### LINKED ISSUE #1529: Audit module
Continued from https://github.com/zotonic/zotonic/issues/1182#issuecomment-261201253.

> What about the amount of time the actions need to be stored? Two months seems reasonable.

Why not forever? In the [event sourcing](http://martinfowler.com/eaaDev/EventSourcing.html) way of building applications, every change is captured and persisted. The idea is to store the events forever to keep a full history of all changes. Why not do the same?

### PR DESCRIPTION
**Pull request by ddeboer**
### Description

Fix #1529.

### Checklist

- [ ] Add schema for `log_audit` table.
- [ ] Update admin templates to show and filter on all `#log_audit{}` properties, similar to the e-mail log.
- [ ] documentation updated
- [x] no BC breaks

### CODE REVIEW

File: modules/mod_audit/mod_audit.erl [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+observe_m_config_update(#m_config_update{module = Module, key = Key, value = Value}, Context) ->
+    log(
+        #log_audit{
+            action = config_update,
+            previous_value = m_config:get(Module, Key),
+            value = Value
+        },
+        Context
+    ).
```
Comment (ddeboer) [inline modules/mod_audit/mod_audit.erl]: All these observers are quite verbose, but I like that it’s very explicit.

File: modules/mod_audit/mod_audit.erl [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+-export([
+    observe_m_config_update/2,
+    observe_module_activate/2,
+    observe_module_deactivate/2,
+    observe_auth_checked/2,
+    observe_auth_logoff/3,
+    observe_edge_insert/2,
+    observe_edge_delete/2,
+    observe_media_update_done/2,
+    observe_rsc_update_done/2
```
Comment (ddeboer) [inline modules/mod_audit/mod_audit.erl]: Any other notifications you’d like to see in the audit log?

**Comment by mworrell** [comment]
Maybe we could hook into (the results of) `mod_ratelimit` and `mod_clamav` ?

**Comment by ddeboer** [comment]
@mworrell Not yet, because they publish no notifications. 😉

**Comment by mworrell** [comment]
It might be an idea to be able to hook into the `z_notifier` to observe the results of `first` or `map` notifications? Wondering what the most efficient method could be.

**Comment by mmzeeman** [comment]
In the module we made it is possible to have an hierarchy of events so you can easily filter on authentication events, resource modification, etc. In our own codebase we have custom events.

What I don't have right now is an admin interface.

**Comment by mworrell** [comment]
See also https://github.com/zotonic/zotonic/pull/2057 for some simple logging of system-events using the current logging module.

**Comment by mworrell** [comment]
This is for an older 0.x - we have now moved to 1.x

I think this could also be done using a Logger module.

**Pull request closed without merging**